js/modules/k6/experimental/streams/readable_streams.go

```diff
@@ -0,0 +1,586 @@
+	// 1. Set stream.[[disturbed]] to true.
+	stream.disturbed = true
+
+	// 2. If stream.[[state]] is "closed", return a promise resolved with undefined.
+	if stream.state == ReadableStreamStateClosed {
+		return newResolvedPromise(stream.vu, goja.Undefined())
+	}
+
+	// 3. If stream.[[state]] is "errored", return a promise rejected with stream.[[storedError]].
+	if stream.state == ReadableStreamStateErrored {
+		return newRejectedPromise(stream.vu, stream.storedError)
+	}
+
+	// 4. Perform ! ReadableStreamClose(stream).
+	stream.close()
+
+	// 5. Let reader be stream.[[reader]].
+	// 6. If reader is not undefined and reader implements ReadableStreamBYOBReader,
+	// Not implemented yet: ReadableStreamBYOBReader is not supported yet.
+
+	// 7. Let sourceCancelPromise be ! stream.[[controller]].[[CancelSteps]](reason).
+	sourceCancelPromise := stream.controller.cancelSteps(reason)
+
+	// 8. Return the result of reacting to sourceCancelPromise with a fulfillment step that returns undefined.
+	promise, err := promiseThen(stream.vu.Runtime(), sourceCancelPromise,
+		// Mimicking Deno's implementation: https://github.com/denoland/deno/blob/main/ext/web/06_streams.js#L405
+		func(goja.Value) {},
+		func(err goja.Value) { throw(stream.vu.Runtime(), err) },
+	)
+	if err != nil {
+		common.Throw(stream.vu.Runtime(), err)
+	}
+
+	return promise
+}
+
+// close implements the specification's [ReadableStreamClose()] abstract operation.
+//
+// [ReadableStreamClose()]: https://streams.spec.whatwg.org/#readable-stream-close
+func (stream *ReadableStream) close() {
+	// 1. Assert: stream.[[state]] is "readable".
+	if stream.state != ReadableStreamStateReadable {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "cannot close a stream that is not readable"))
+	}
+
+	// 2. Set stream.[[state]] to "closed".
+	stream.state = ReadableStreamStateClosed
+
+	// 3. Let reader be stream.[[reader]].
+	reader := stream.reader
+
+	// 4. If reader is undefined, return.
+	if reader == nil {
+		return
+	}
+
+	// 5. Resolve reader.[[closedPromise]] with undefined.
+	genericReader, ok := reader.(ReadableStreamGenericReader)
+	if !ok {
+		common.Throw(stream.vu.Runtime(), newError(AssertionError, "reader is not a ReadableStreamGenericReader"))
+	}
+
+	_, resolveFunc, _ := genericReader.GetClosed()
+	resolveFunc(goja.Undefined())
+
+	// 6. If reader implements ReadableStreamDefaultReader,
+	defaultReader, ok := reader.(*ReadableStreamDefaultReader)
+	if ok {
+		// 6.1. Let readRequests be reader.[[readRequests]].
+		readRequests := defaultReader.readRequests
+
+		// 6.2. Set reader.[[readRequests]] to an empty list.
+		defaultReader.readRequests = []ReadRequest{}
+
+		// 6.3. For each readRequest of readRequests,
+		for _, readRequest := range readRequests {
+			// 6.3.1. Perform readRequest’s close steps.
+			stream.vu.RegisterCallback()(func() error {
+				readRequest.closeSteps()
+				return nil
+			})
+		}
```

Comment:
I am a bit confused - why are we doing this in a callback? 

Also you are hitting the usual range gotcha that is fixed in go1.22+ (with the appropriate go.mod setting).

Reply:
Good catch! I think the reason why I left this here is because at some point, `closeSteps` involved a goroutine call, so I decided to use callbacks to ensure their execution as part of the event loop. But yeah, it's no longer needed, I think.